### ISSUE TITLE
[ClientBootstrap]: segfault at program exit

### ISSUE THREAD

**Issue by grrtrr** [awslabs/aws-crt-cpp#388]
### Summary

This issue happens regardless of whether `ClientBootstrap::EnableBlockingShutdown()` has been called, and it is caused by returning from `~ClientBootstrap()` while asynchronous shutdown is still in progress.

In production, we are repeatedly getting the same backtrace:
```console
FATAL: Received signal 11 (Segmentation fault)
Backtrace (most recent call first)
#6	 <?> at 0x564a57ceca28 in Aws::Utils::Logging::s_aws_logger_redirect_log()
#5	 <?> at 0x564a57d4bb2b in aws_default_dns_resolve
#4	 <?> at 0x564a57d4dfed in resolver_thread_fn
#3	 <?> at 0x564a57d5a117 in thread_fn
#2	 <?> at 0x564a5753f32c in thread_metrics_start_routine
#1	 <?> at 0x7f04dcfb86db in start_thread
```

### Details

The orderly shutdown of the client bootstrap is invoked by calling `CleanupCrt()`, followed by shutting down CRT logging (see 
https://github.com/aws/aws-sdk-cpp/issues/1995).
```c++
// aws core/source/Globals.cpp
    void CleanupCrt()
    {
        Aws::SetDefaultClientBootstrap(nullptr);
        // ...
    }
```

The `s_client_bootstrap_destroy_impl` invokes the "blocking shutdown" callback right after the `aws_event_loop_group_release` and `aws_host_resolver_release` calls:
```c
static void s_client_bootstrap_destroy_impl(struct aws_client_bootstrap *bootstrap) {
    AWS_ASSERT(bootstrap);
    AWS_LOGF_DEBUG(AWS_LS_IO_CHANNEL_BOOTSTRAP, "id=%p: destroying", (void *)bootstrap);
    aws_client_bootstrap_shutdown_complete_fn *on_shutdown_complete = bootstrap->on_shutdown_complete;
    void *user_data = bootstrap->user_data;

    aws_event_loop_group_release(bootstrap->event_loop_group);
    aws_host_resolver_release(bootstrap->host_resolver);

    if (on_shutdown_complete) {
        on_shutdown_complete(user_data);
    }
}
```

However, both the `event_loop_group` and the `host_resolver` perform _asynchronous shutdown_, i.e. their "destroy" functions merely trigger the shutdown of threads that are still running; they do not await thread exit.

The _problem_ now arises that the `event_loop_group` and clearly the `host_resolver` still issue calls to the CRT logging system.
We can not afford to shut down CRT logging (via `Aws::Utils::Logging::ShutdownCRTLogging()`) while these threads are still exiting,
otherwise the above segfault will result.

### Possible fix

Populate the shutdown callback functions of both the contained `event_loop_group` and `host_resolver`.
Have them set a flag/counter to indicate that the shutdown callbacks (at thread exit) have been invoked.

For good measure, also have the counter set in the shutdown of the `aws_client_bootstrap`.

In the destructor of `ClientBootstrap`, await, via condition variable, all 3 shutdowns.

Ignore the "blocking shutdown" boolean variable (due to the above described problems).
Use a _timeout_ for the condition variable, in case an exiting thread failed to invoke the final callback.
This is to avoid a _hanging program_ at end (the current solution uses a `std::promise`, which would block indefinitely if `set_value` was not called for an unexpected reason).

**Comment by bretambrose** [comment]
I don't quite follow what you mean by blocking shutdown as the async callbacks are only intended as individual notices of cleanup completion (for a single object).

The CRT has a shutdown option (https://github.com/awslabs/aws-crt-cpp/blob/main/include/aws/crt/Api.h#L38) which will wait for all CRT-spawned threads to complete (https://github.com/awslabs/aws-crt-cpp/blob/main/source/Api.cpp#L76).  You can enabled it as a flag on the ApiHandle.  If the C++ SDK is not giving that option then it seems like an issue there not here.

**Comment by bretambrose** [comment]
On followup, I see what the blocking shutdown thing is in ClientBootstrap.  Yeah, that shouldn't be used and doesn't do anything particularly useful.

**Comment by grrtrr** [comment]
Oh I see the API has changed. So maybe this is fixed in `v0.17.34`?
We observed this repeatedly on `v0.17.1`.

**Comment by bretambrose** [comment]
My vague recollection is that the C++ SDK may have disabled blocking shutdown.  The managed thread join has existed for about a year and half now: https://github.com/awslabs/aws-crt-cpp/releases/tag/v0.11.0

**Comment by grrtrr** [comment]
@bretambrose - you are right, the shutdown of the `host_resolver` threads is done in `~ApiHandle()`, as the thread type is 'managed'.
We are still seeing this, even after #389. My guess is that it tries to log a data structure that has been freed at shutdown, and hence crashes. Will open a new issue if I find such a case.
